### PR TITLE
use pdqselect for median_algorithm selection

### LINKED ISSUE #2926: Use a selection algorithm instead of a sorting algorithm in median algorithm
The current code uses Rust’s sort routines, which are O(n log n), as opposed to a proper selection algorithm, which is O(n).

This is very low priority, as I cannot see this being a bottleneck in any practical scenario.  I suspect that unless such an algorithm is added to the standard library, this issue is probably not worth fixing.

### PR DESCRIPTION
**Pull request by expenses**
This would resolve #2926, but `median_algorithm` isn't currently used. I haven't written a test yet, but I'd like to if it's important enough.

### CODE REVIEW

**Comment by gavofyork** [comment]
> This would resolve #2926 

would it?

https://docs.rs/pdqselect/0.1.0/pdqselect/fn.select.html

> "This sort is in-place, unstable, and O(n log n) worst-case"

**Comment by expenses** [comment]
I ran a simple benchmark comparing getting the median with both `select` and `unsafe_sort` [here](https://gist.github.com/expenses/d2d84305aeaa512200676dc0198e9f81#file-bench-rs). Results:

```
test bench_select ... bench:       1,898 ns/iter (+/- 208)
test bench_sort   ... bench:       9,288 ns/iter (+/- 1,340)
```

**Comment by rphmeier** [comment]
Thanks! Mind if I ice this until #3583 is in, since it contains a much larger change-set on BABE that this might conflict with?

**Comment by expenses** [comment]
Of course!

File: core/consensus/babe/src/lib.rs [hunk 1]
```diff
@@ -110,6 +110,7 @@ use slots::{CheckedHeader, check_equivocation};
 use futures::prelude::*;
 use futures01::Stream as _;
 use log::{error, warn, debug, info, trace};
+use pdqselect::select;
```
Comment (andresilva) [inline core/consensus/babe/src/lib.rs]:
```suggestion
```

File: core/consensus/babe/src/lib.rs [hunk 2]
```diff
@@ -717,8 +718,8 @@ fn median_algorithm(
 			t + Duration::new(secs, nanos)
 		}).collect();
 
-		// FIXME #2926: use a selection algorithm instead of a full sorting algorithm.
-		new_list.sort_unstable();
+		// Use a partial sort to move the median timestamp to the middle of the list
+		select(&mut new_list, num_timestamps / 2);
```
Comment (andresilva) [inline core/consensus/babe/src/lib.rs]:
```suggestion
		pdqselect::select(&mut new_list, num_timestamps / 2);
```

Comment (andresilva) [reply core/consensus/babe/src/lib.rs]: nitpick: maybe easier to understand what this is just by reading this line alone. feel free to ignore it though :)

Comment (rphmeier) [reply core/consensus/babe/src/lib.rs]: yes, i find this style much better for free functions - it means that whenever you see a function call without a module prefix, that it's defined in the same file as you're reading right now.

**Comment by rphmeier** [comment]
Thanks!

**Pull request merged**